File: src/main/store.js
const fs = require('fs');
const path = require('path');

class Store {
  constructor(dataPath) {
    this.filePath = path.join(dataPath, 'tasks.json');
    this.data = this._load();
    this._migrate();
  }

  /** 尝试从旧路径迁移数据到新路径。支持多个候选来源，picks newest. */
  static migrateFromLegacy(legacyCandidates, newPath) {
    const newFile = path.join(newPath, 'tasks.json');
    const candidates = Array.isArray(legacyCandidates) ? legacyCandidates : [legacyCandidates];
    try {
      if (fs.existsSync(newFile)) return; // already migrated
      // pick the newest legacy file that exists
      let bestSrc = null;
      let bestMtime = 0;
      for (const c of candidates) {
        const f = path.join(c, 'tasks.json');
        if (fs.existsSync(f)) {
          const m = fs.statSync(f).mtimeMs;
          if (m > bestMtime) { bestMtime = m; bestSrc = f; }
        }
      }
      if (bestSrc) {
        if (!fs.existsSync(newPath)) fs.mkdirSync(newPath, { recursive: true });
        fs.copyFileSync(bestSrc, newFile);
        console.log('[Store] Migrated data from', bestSrc, 'to', newFile);
      }
    } catch (e) {
      console.error('[Store] Migration failed:', e);
    }
  }

  _load() {
    try {
      if (fs.existsSync(this.filePath)) {
        const raw = fs.readFileSync(this.filePath, 'utf-8');
        return JSON.parse(raw);
      }
    } catch (e) {
      console.error('[Store] Failed to load:', e);
    }
    return {
      tasks: [], nextId: 1,
      tags: {}, snippets: [], nextSnippetId: 1,
      settings: { cornerTrigger: 'top-left' },
    };
  }

  _migrate() {
    let changed = false;
    // tags: Array → Object { name: { color } }
    if (Array.isArray(this.data.tags)) {
      const obj = {};
      for (const t of this.data.tags) {
        obj[t] = { color: null };
      }
      this.data.tags = obj;
      changed = true;
    }
    if (!this.data.tags) { this.data.tags = {}; changed = true; }
    if (!this.data.snippets) { this.data.snippets = []; changed = true; }
    if (!this.data.nextSnippetId) { this.data.nextSnippetId = 1; changed = true; }
    if (!this.data.settings) { this.data.settings = {}; changed = true; }
    if (!this.data.settings.cornerTrigger) { this.data.settings.cornerTrigger = 'top-left'; changed = true; }

    for (const task of this.data.tasks) {
      if (!task.priority) { task.priority = 'none'; changed = true; }
      if (!task.listId) { task.listId = 'inbox'; changed = true; }
      if (!task.description && task.description !== '') { task.description = ''; changed = true; }
      if (!task.tags) { task.tags = []; changed = true; }
      if (!task.subtasks) { task.subtasks = []; changed = true; }
      if (task.dueDate === undefined) { task.dueDate = null; changed = true; }
      if (!task.nextSubId) { task.nextSubId = 1; changed = true; }
    }
    if (changed) this._save();
  }

  _save() {
    const dir = path.dirname(this.filePath);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(this.filePath, JSON.stringify(this.data, null, 2), 'utf-8');
  }

  getDataPath() {
    return this.filePath;
  }

  // ======= Tasks =======

  getTasks() {
    return this.data.tasks;
  }

  getTask(id) {
    return this.data.tasks.find(t => t.id === id) || null;
  }

  addTask(title, priority = 'none', tags = [], dueDate = null) {
    const task = {
      id: this.data.nextId++,
      title: title.trim(),
      description: '',
      completed: false,
      priority,
      tags,
      subtasks: [],
      nextSubId: 1,
      dueDate,
      listId: 'inbox',
      createdAt: new Date().toISOString(),
      completedAt: null,
    };
    this.data.tasks.unshift(task);
    this._syncGlobalTags(tags);
    this._save();
    return task;
  }

  toggleTask(id) {
    const task = this.data.tasks.find(t => t.id === id);
    if (task) {
      task.completed = !task.completed;
      task.completedAt = task.completed ? new Date().toISOString() : null;
      this._save();
    }
    return task;
  }

  deleteTask(id) {
    this.data.tasks = this.data.tasks.filter(t => t.id !== id);
    this._save();
  }

  updateTask(id, updates) {
    const task = this.data.tasks.find(t => t.id === id);
    if (task) {
      if (updates.tags) this._syncGlobalTags(updates.tags);
      Object.assign(task, updates);
      this._save();
    }
    return task;
  }

  clearCompleted() {
    this.data.tasks = this.data.tasks.filter(t => !t.completed);
    this._save();
  }

  // ======= Subtasks =======

  addSubtask(taskId, title) {
    const task = this.data.tasks.find(t => t.id === taskId);
    if (!task) return null;
    const sub = {
      id: task.nextSubId++,
      title: title.trim(),
      completed: false,
    };
    task.subtasks.push(sub);
    this._save();
    return sub;
  }

  toggleSubtask(taskId, subId) {
    const task = this.data.tasks.find(t => t.id === taskId);
    if (!task) return null;
    const sub = task.subtasks.find(s => s.id === subId);
    if (sub) {
      sub.completed = !sub.completed;
      this._save();
    }
    return task;
  }

  deleteSubtask(taskId, subId) {
    const task = this.data.tasks.find(t => t.id === taskId);
    if (!task) return null;
    task.subtasks = task.subtasks.filter(s => s.id !== subId);
    this._save();
    return task;
  }

  updateSubtask(taskId, subId, title) {
    const task = this.data.tasks.find(t => t.id === taskId);
    if (!task) return null;
    const sub = task.subtasks.find(s => s.id === subId);
    if (sub) {
      sub.title = title.trim();
      this._save();
    }
    return task;
  }

  // ======= Tags (Object: { tagName: { color } }) =======

  getAllTags() {
    return this.data.tags;
  }

  setTagColor(tagName, color) {
    if (!this.data.tags[tagName]) {
      this.data.tags[tagName] = { color };
    } else {
      this.data.tags[tagName].color = color;
    }
    this._save();
    return this.data.tags;
  }

  _syncGlobalTags(tags) {
    for (const tag of tags) {
      if (!this.data.tags[tag]) {
        this.data.tags[tag] = { color: null };
      }
    }
  }

  deleteTag(tagName) {
    delete this.data.tags[tagName];
    for (const task of this.data.tasks) {
      task.tags = task.tags.filter(t => t !== tagName);
    }
    this._save();
  }

  // ======= Snippets (快捷粘贴) =======

  getSnippets() {
    return this.data.snippets;
  }

  addSnippet(label, value, type = 'text') {
    const s = {
      id: this.data.nextSnippetId++,
      label: (label || '').trim(),
      value: value || '',
      type, // 'text' | 'password'
      createdAt: new Date().toISOString(),
    };
    this.data.snippets.push(s);
    this._save();
    return s;
  }

  updateSnippet(id, updates) {
    const s = this.data.snippets.find(x => x.id === id);
    if (s) {
      Object.assign(s, updates);
      this._save();
    }
    return s;
  }

  deleteSnippet(id) {
    this.data.snippets = this.data.snippets.filter(s => s.id !== id);
    this._save();
  }

  reorderSnippet(id, targetIndex) {
    const idx = this.data.snippets.findIndex(s => s.id === id);
    if (idx === -1) return;
    const [s] = this.data.snippets.splice(idx, 1);
    this.data.snippets.splice(targetIndex, 0, s);
    this._save();
  }

  // ======= Settings =======

  getSettings() {
    return this.data.settings;
  }

  updateSettings(updates) {
    Object.assign(this.data.settings, updates);
    this._save();
    return this.data.settings;
  }
}

module.exports = Store;

File: NookSwift/Nook/Core/Store.swift
import Foundation
import SwiftUI

struct TagInfo: Codable, Equatable {
    var color: String?
}

struct StoreData: Codable {
    var tasks: [NookTask]
    var nextId: Int
    var tags: [String: TagInfo]
    var snippets: [Snippet]
    var nextSnippetId: Int
    var settings: NookSettings

    init() {
        tasks = []
        nextId = 1
        tags = [:]
        snippets = []
        nextSnippetId = 1
        settings = NookSettings()
    }
}

@MainActor
final class Store: ObservableObject {
    @Published var tasks: [NookTask] = []
    @Published var tags: [String: TagInfo] = [:]
    @Published var snippets: [Snippet] = []
    @Published var settings: NookSettings = NookSettings()
    @Published var confettiToken: UUID?

    private var nextId: Int = 1
    private var nextSnippetId: Int = 1
    private let filePath: URL

    // External-write watcher state
    private var fileMonitor: DispatchSourceFileSystemObject?
    private var watchedFD: Int32 = -1
    /// Suppress reload reactions to our own writes — extend on each save().
    private var ignoreReloadUntil: Date = .distantPast

    static let tagColors: [String] = [
        "#FF6B6B", "#FF8A65", "#FFB74D", "#FFD54F", "#FFF176",
        "#DCE775", "#AED581", "#81C784", "#4DB6AC", "#4DD0E1",
        "#4FC3F7", "#64B5F6", "#7986CB", "#9575CD", "#BA68C8",
        "#CE93D8", "#F06292", "#E57373", "#A1887F", "#90A4AE",
    ]

    init() {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dataDir = appSupport.appendingPathComponent("Nook/data")
        try? FileManager.default.createDirectory(at: dataDir, withIntermediateDirectories: true)
        filePath = dataDir.appendingPathComponent("tasks.json")
        let isFirstRun = !FileManager.default.fileExists(atPath: filePath.path)
        load()
        if isFirstRun {
            seedWelcomeTask()
        }
        startWatching()
    }

    /// First-run welcome task + starter snippets — the snippets section at the bottom
    /// holds the actual install command and AI prompts (one-click copy), so the
    /// welcome task can stay short and explain the WHAT, not the HOW.
    private func seedWelcomeTask() {
        let welcomeDesc = """
        Let your AI agent (Claude, Cursor, etc.) drop tasks straight into Nook — no need to switch back to this panel and type.

        ✨ How it works
        Have your AI install the nooktodo CLI once (the command is in Snippets at the bottom). After that, just tell the AI "add a task: X" — it writes to Nook, and this panel updates live.

        📝 Example
        You: "Add a task: buy milk tomorrow, high priority"
        AI runs: nooktodo "buy milk" -p high -d tomorrow
        ↓
        Task appears here instantly ✓

        Tick the subtasks below as you go, then delete this whole task 👋
        """
        let task = NookTask(
            id: 1,
            title: "👋 Let an AI add tasks for you",
            description: welcomeDesc,
            completed: false,
            priority: .none,
            tags: [],
            subtasks: [
                Subtask(id: 1, title: "Copy the [Install nooktodo] snippet from the bottom rail", completed: false),
                Subtask(id: 2, title: "Ask any AI agent (Claude / Cursor) to run that command", completed: false),
                Subtask(id: 3, title: "Tell the AI \"add a task: XX\" and watch it appear here", completed: false),
            ],
            nextSubId: 4,
            dueDate: nil,
            listId: "inbox",
            createdAt: ISO8601DateFormatter().string(from: Date()),
            completedAt: nil
        )
        tasks = [task]
        nextId = 2

        // Starter snippets — give users one-click access to install + AI prompts
        let installCmd = #"NOOK=$(find /Applications ~/Applications -name Nook.app -maxdepth 2 2>/dev/null | head -1) && mkdir -p ~/.local/bin && ln -sf "$NOOK/Contents/Resources/nooktodo" ~/.local/bin/nooktodo && echo "✓ nooktodo installed"#
        let now = ISO8601DateFormatter().string(from: Date())
        snippets = [
            Snippet(id: 1, label: "📦 Install nooktodo (have AI run this)",
                    value: installCmd, type: "command", createdAt: now),
            Snippet(id: 2, label: "💡 AI prompt template",
                    value: "Add a task: X, priority high, due tomorrow",
                    type: "text", createdAt: now),
            Snippet(id: 3, label: "📝 nooktodo command format",
                    value: #"nooktodo "title" -t tag -p high -d tomorrow"#,
                    type: "command", createdAt: now),
        ]
        nextSnippetId = 4

        save()
    }

    deinit {
        fileMonitor?.cancel()
        if watchedFD >= 0 { close(watchedFD) }
    }

    /// Watch the data directory for changes — atomic writes swap inodes, so watching
    /// the file directly would lose the descriptor on every save. The directory
    /// vnode is stable across the rename. Reload whenever any change fires.
    private func startWatching() {
        let dir = filePath.deletingLastPathComponent()
        watchedFD = open(dir.path, O_EVTONLY)
        guard watchedFD >= 0 else { return }
        let source = DispatchSource.makeFileSystemObjectSource(
            fileDescriptor: watchedFD,
            eventMask: [.write, .extend, .rename, .delete],
            queue: .main
        )
        source.setEventHandler { [weak self] in
            self?.reloadIfExternalChange()
        }
        source.setCancelHandler { [weak self] in
            if let fd = self?.watchedFD, fd >= 0 {
                close(fd)
                self?.watchedFD = -1
            }
        }
        source.resume()
        fileMonitor = source
    }

    private func reloadIfExternalChange() {
        // Skip reactions to our own recent saves.
        guard Date() > ignoreReloadUntil else { return }
        load()
    }

    private func load() {
        guard FileManager.default.fileExists(atPath: filePath.path) else { return }
        do {
            let data = try Data(contentsOf: filePath)
            let decoded = try JSONDecoder().decode(StoreData.self, from: data)
            tasks = decoded.tasks
            nextId = decoded.nextId
            tags = decoded.tags
            snippets = decoded.snippets
            nextSnippetId = decoded.nextSnippetId
            settings = decoded.settings
        } catch {
            print("[Store] Failed to load: \(error)")
        }
    }

    private func save() {
        let storeData = StoreData(
            tasks: tasks, nextId: nextId,
            tags: tags, snippets: snippets,
            nextSnippetId: nextSnippetId, settings: settings
        )
        // Suppress reload reactions to our own write — atomic write fires multiple
        // dir events (temp create, rename, delete temp).
        ignoreReloadUntil = Date().addingTimeInterval(0.6)
        do {
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
            let data = try encoder.encode(storeData)
            try data.write(to: filePath, options: .atomic)
        } catch {
            print("[Store] Failed to save: \(error)")
        }
    }

    // MARK: - Tasks

    @discardableResult
    func addTask(title: String, description: String = "", priority: NookTask.Priority = .none, tags: [String] = [], dueDate: String? = nil) -> Int {
        let taskId = nextId
        let task = NookTask(
            id: taskId, title: title.trimmingCharacters(in: .whitespaces),
            description: description, completed: false, priority: priority,
            tags: tags, subtasks: [], nextSubId: 1,
            dueDate: dueDate, listId: "inbox",
            createdAt: ISO8601DateFormatter().string(from: Date()), completedAt: nil
        )
        nextId += 1
        tasks.insert(task, at: 0)
        syncGlobalTags(tags)
        save()
        return taskId
    }

    func toggleTask(_ id: Int) {
        guard let idx = tasks.firstIndex(where: { $0.id == id }) else { return }
        let wasCompleted = tasks[idx].completed
        tasks[idx].completed.toggle()
        tasks[idx].completedAt = tasks[idx].completed ? ISO8601DateFormatter().string(from: Date()) : nil
        if !wasCompleted {
            confettiToken = UUID()
        }
        save()
    }

    func deleteTask(_ id: Int) {
        tasks.removeAll { $0.id == id }
        save()
    }

    func updateTask(_ id: Int, title: String? = nil, description: String? = nil,
                    priority: NookTask.Priority? = nil, tags: [String]? = nil,
                    dueDate: String?? = nil) {
        guard let idx = tasks.firstIndex(where: { $0.id == id }) else { return }
        if let t = title { tasks[idx].title = t }
        if let d = description { tasks[idx].description = d }
        if let p = priority { tasks[idx].priority = p }
        if let tgs = tags {
            tasks[idx].tags = tgs
            syncGlobalTags(tgs)
        }
        if let dd = dueDate { tasks[idx].dueDate = dd }
        save()
    }

    func reorderTask(_ id: Int, to targetIndex: Int) {
        guard let idx = tasks.firstIndex(where: { $0.id == id }) else { return }
        let task = tasks.remove(at: idx)
        let insertAt = min(targetIndex, tasks.count)
        tasks.insert(task, at: insertAt)
        save()
    }

    func clearCompleted() {
        tasks.removeAll { $0.completed }
        save()
    }

    // MARK: - Subtasks

    func addSubtask(_ taskId: Int, title: String) {
        guard let idx = tasks.firstIndex(where: { $0.id == taskId }) else { return }
        let sub = Subtask(id: tasks[idx].nextSubId, title: title.trimmingCharacters(in: .whitespaces), completed: false)
        tasks[idx].nextSubId += 1
        tasks[idx].subtasks.append(sub)
        save()
    }

    func toggleSubtask(_ taskId: Int, subId: Int) {
        guard let idx = tasks.firstIndex(where: { $0.id == taskId }),
              let si = tasks[idx].subtasks.firstIndex(where: { $0.id == subId }) else { return }
        tasks[idx].subtasks[si].completed.toggle()
        save()
    }

    func deleteSubtask(_ taskId: Int, subId: Int) {
        guard let idx = tasks.firstIndex(where: { $0.id == taskId }) else { return }
        tasks[idx].subtasks.removeAll { $0.id == subId }
        save()
    }

    func updateSubtask(_ taskId: Int, subId: Int, title: String) {
        guard let idx = tasks.firstIndex(where: { $0.id == taskId }),
              let si = tasks[idx].subtasks.firstIndex(where: { $0.id == subId }) else { return }
        tasks[idx].subtasks[si].title = title.trimmingCharacters(in: .whitespaces)
        save()
    }

    // MARK: - Tags

    func setTagColor(_ name: String, color: String?) {
        tags[name] = TagInfo(color: color)
        save()
    }

    func deleteTag(_ name: String) {
        tags.removeValue(forKey: name)
        for i in tasks.indices {
            tasks[i].tags.removeAll { $0 == name }
        }
        save()
    }

    private func syncGlobalTags(_ tagNames: [String]) {
        for name in tagNames where tags[name] == nil {
            tags[name] = TagInfo(color: nil)
        }
    }

    // MARK: - Snippets

    func addSnippet(label: String, value: String, type: String = "text") {
        let s = Snippet(
            id: nextSnippetId,
            label: label.trimmingCharacters(in: .whitespaces),
            value: value, type: type,
            createdAt: ISO8601DateFormatter().string(from: Date())
        )
        nextSnippetId += 1
        snippets.append(s)
        save()
    }

    func updateSnippet(_ id: Int, label: String? = nil, value: String? = nil) {
        guard let idx = snippets.firstIndex(where: { $0.id == id }) else { return }
        if let l = label { snippets[idx].label = l }
        if let v = value { snippets[idx].value = v }
        save()
    }

    func deleteSnippet(_ id: Int) {
        snippets.removeAll { $0.id == id }
        save()
    }

    // MARK: - Settings

    func updateSettings(_ newSettings: NookSettings) {
        settings = newSettings
        save()
    }

    func updateCorner(_ corner: NookSettings.CornerTrigger) {
        settings.cornerTrigger = corner
        save()
    }

    func updatePanelSize(heightRatio: Double, width: Double) {
        settings.panelHeightRatio = heightRatio
        settings.panelWidth = width
        save()
    }
}

extension StoreData {
    init(tasks: [NookTask], nextId: Int, tags: [String: TagInfo],
         snippets: [Snippet], nextSnippetId: Int, settings: NookSettings) {
        self.tasks = tasks
        self.nextId = nextId
        self.tags = tags
        self.snippets = snippets
        self.nextSnippetId = nextSnippetId
        self.settings = settings
    }
}
